File: src/java/non-corba/com/metricstream/omg/naming/NamingContext.java
package com.metricstream.omg.naming;

/**
 * Interface for the naming context, which is a container for name-to-object bindings.
 * This is the non-CORBA equivalent of org.omg.CosNaming.NamingContext.
 */
public interface NamingContext {

    /**
     * Binds a name to an object in this naming context.
     * 
     * @param n The name to bind
     * @param obj The object to bind
     * @throws NotFound If the name contains a component that does not exist
     * @throws CannotProceed If the implementation cannot proceed with the operation
     * @throws InvalidName If the name is invalid
     * @throws AlreadyBound If the name is already bound
     */
    void bind(Name n, Object obj) throws NotFound, CannotProceed, InvalidName, AlreadyBound;

    /**
     * Binds a name to an object in this naming context, even if the name is already bound.
     * 
     * @param n The name to bind
     * @param obj The object to bind
     * @throws NotFound If the name contains a component that does not exist
     * @throws CannotProceed If the implementation cannot proceed with the operation
     * @throws InvalidName If the name is invalid
     */
    void rebind(Name n, Object obj) throws NotFound, CannotProceed, InvalidName;

    /**
     * Creates a new naming context and binds it to the specified name.
     * 
     * @param n The name to bind
     * @return The newly created naming context
     * @throws NotFound If the name contains a component that does not exist
     * @throws CannotProceed If the implementation cannot proceed with the operation
     * @throws InvalidName If the name is invalid
     * @throws AlreadyBound If the name is already bound
     */
    NamingContext bind_new_context(Name n) throws NotFound, CannotProceed, InvalidName, AlreadyBound;

    /**
     * Resolves a name to an object.
     * 
     * @param n The name to resolve
     * @return The object bound to the name
     * @throws NotFound If the name cannot be found
     * @throws CannotProceed If the implementation cannot proceed with the operation
     * @throws InvalidName If the name is invalid
     */
    Object resolve(Name n) throws NotFound, CannotProceed, InvalidName;

    /**
     * Unbinds a name from an object.
     * 
     * @param n The name to unbind
     * @throws NotFound If the name cannot be found
     * @throws CannotProceed If the implementation cannot proceed with the operation
     * @throws InvalidName If the name is invalid
     */
    void unbind(Name n) throws NotFound, CannotProceed, InvalidName;

    /**
     * Lists the bindings in this naming context.
     * 
     * @param how_many The maximum number of bindings to return
     * @param bl The binding list to fill
     * @param bi The binding iterator to fill
     */
    void list(int how_many, BindingListHolder bl, BindingIteratorHolder bi);

    /**
     * Destroys this naming context.
     * 
     * @throws NotEmpty If the naming context is not empty
     */
    void destroy() throws NotEmpty;
}

File: src/main/kotlin/com/metricstream/naming/NamingContext.kt
package com.metricstream.naming

/**
 * Interface for a naming context.
 * Equivalent to CosNaming::NamingContext in CORBA.
 */
interface NamingContext {
    /**
     * Binds a name to an object.
     * 
     * @param name The name to bind
     * @param obj The object to bind
     * @throws NotFound if a component of the name does not exist
     * @throws CannotProceed if the implementation cannot proceed
     * @throws InvalidName if the name is invalid
     * @throws AlreadyBound if the name is already bound
     */
    @Throws(NotFound::class, CannotProceed::class, InvalidName::class, AlreadyBound::class)
    fun bind(name: Name, obj: Any)
    
    /**
     * Binds a name to a naming context.
     * 
     * @param name The name to bind
     * @param context The naming context to bind
     * @throws NotFound if a component of the name does not exist
     * @throws CannotProceed if the implementation cannot proceed
     * @throws InvalidName if the name is invalid
     * @throws AlreadyBound if the name is already bound
     */
    @Throws(NotFound::class, CannotProceed::class, InvalidName::class, AlreadyBound::class)
    fun bind_context(name: Name, context: NamingContext)
    
    /**
     * Rebinds a name to an object.
     * 
     * @param name The name to rebind
     * @param obj The object to rebind
     * @throws NotFound if a component of the name does not exist
     * @throws CannotProceed if the implementation cannot proceed
     * @throws InvalidName if the name is invalid
     */
    @Throws(NotFound::class, CannotProceed::class, InvalidName::class)
    fun rebind(name: Name, obj: Any)
    
    /**
     * Rebinds a name to a naming context.
     * 
     * @param name The name to rebind
     * @param context The naming context to rebind
     * @throws NotFound if a component of the name does not exist
     * @throws CannotProceed if the implementation cannot proceed
     * @throws InvalidName if the name is invalid
     */
    @Throws(NotFound::class, CannotProceed::class, InvalidName::class)
    fun rebind_context(name: Name, context: NamingContext)
    
    /**
     * Resolves a name to an object.
     * 
     * @param name The name to resolve
     * @return The object bound to the name
     * @throws NotFound if the name does not exist
     * @throws CannotProceed if the implementation cannot proceed
     * @throws InvalidName if the name is invalid
     */
    @Throws(NotFound::class, CannotProceed::class, InvalidName::class)
    fun resolve(name: Name): Any
    
    /**
     * Unbinds a name.
     * 
     * @param name The name to unbind
     * @throws NotFound if the name does not exist
     * @throws CannotProceed if the implementation cannot proceed
     * @throws InvalidName if the name is invalid
     */
    @Throws(NotFound::class, CannotProceed::class, InvalidName::class)
    fun unbind(name: Name)
    
    /**
     * Creates a new context and binds it to the specified name.
     * 
     * @param name The name to bind
     * @return The new naming context
     * @throws NotFound if a component of the name does not exist
     * @throws CannotProceed if the implementation cannot proceed
     * @throws InvalidName if the name is invalid
     * @throws AlreadyBound if the name is already bound
     */
    @Throws(NotFound::class, CannotProceed::class, InvalidName::class, AlreadyBound::class)
    fun bind_new_context(name: Name): NamingContext
    
    /**
     * Destroys this naming context.
     * 
     * @throws NotEmpty if the context is not empty
     */
    @Throws(NotEmpty::class)
    fun destroy()
    
    /**
     * Lists the bindings in this naming context.
     * 
     * @param howMany The maximum number of bindings to return
     * @param bindingList The list to fill with bindings
     * @param iterator An iterator for the remaining bindings
     * @return true if there are more bindings
     */
    fun list(howMany: Int, bindingList: BindingList): BindingIterator?
}

/**
 * Interface for iterating over bindings in a naming context.
 * Equivalent to CosNaming::BindingIterator in CORBA.
 */
interface BindingIterator {
    /**
     * Returns the next binding.
     * 
     * @return The next binding, or null if there are no more bindings
     */
    fun next_one(): Binding?
    
    /**
     * Returns the next n bindings.
     * 
     * @param howMany The maximum number of bindings to return
     * @param bindingList The list to fill with bindings
     * @return true if there are more bindings
     */
    fun next_n(howMany: Int, bindingList: BindingList): Boolean
    
    /**
     * Destroys this iterator.
     */
    fun destroy()
}
